metrics.py

```diff
@@ -443,72 +459,176 @@ def __init__(self, exogenous_attr: Collection):
+            small_eigen_vals_indices = np.where(np.abs(eigen_vals) < thr)[0]
+            if small_eigen_vals_indices.size == 0:
+                is_linearly_dependent = False
+            else:
+                # For each linear dependency (corresponding to an eigen vector with a low value), mark the indices of
+                # features that are involved (identified by a non-zero coefficient in the eigen vector).
+                dependent_features_flags = np.zeros((small_eigen_vals_indices.size, self.num_of_features))
+                for i in range(small_eigen_vals_indices.size):
+                    dependent_features_flags[
+                        i, np.where(np.abs(eigen_vecs[:, small_eigen_vals_indices[i]]) > thr)[0]] = 1
+
+                # Calculate the fraction of dependencies each feature is involved in.
+                fraction_of_dependencies_involved = dependent_features_flags.mean(axis=0)
+
+                # Sort the features (their indices) by the fraction of dependencies they are involved in (remove first
+                # features that are involved in more dependencies). Break ties by the original order of the features in
+                # the array (for the consistency of sorting. E.g, if we need to get rid of degree features, always
+                # remove them by the order of nodes).
+                removal_order = np.lexsort((np.arange(self.num_of_features), -fraction_of_dependencies_involved))
+
+                # Iterate the metrics to find one with multiple features, namely effective number of features that is
+                # larger than 1 ('trimmable'). We prefer to trim metrics rather than totally eliminate them from the
+                # collection.
+                i = 0
+                cur_metric = self.get_metric_by_feat_idx(removal_order[i])
+                is_trimmable = cur_metric._get_effective_feature_count() > 1
+                while (not is_trimmable and i < removal_order.size - 1 and fraction_of_dependencies_involved[
+                    removal_order[i]] > 0):
+                    i += 1
+                    cur_metric = self.get_metric_by_feat_idx(removal_order[i])
+                    is_trimmable = cur_metric._get_effective_feature_count() > 1
+
+                # If a trimmable metric was not found (i.e., all features that are involved in the dependency are of
+                # metrics with an effective number of features of 1), totally remove the metric that is involved in most
+                # dependencies.
+                if not is_trimmable:
+                    first_metric = self.get_metric_by_feat_idx(removal_order[0])
+                    print(f"Removing the metric {str(first_metric)} from the collection to fix multi-collinearity")
+                    self.metrics = tuple([m for m in self.metrics if str(m) != str(first_metric)])
+                    self.requires_graph = any([x.requires_graph for x in self.metrics])
```

Comment:
The fact that we need to update self.requires_graph here is, in my opinion, an indication of a bad design on our behalf - It's happening because there are two attributes (MetricsCollection.metrics and MetricsCollection.requires_graph) that depend on each other, but their implementation is completely independent.

What we should have, is `collection.remove_metric(idx)` or something, which would update requires_graph behind the scenes.

It's not so bad and not sure we need to fix this, but in the future when we'll have more attributes like `requires_graph`, it might be a problem.

Reply:
I agree it's not ideal, but actually we have more attributes like `requires_graph` (e.g. `is_sparse` and 'metric_names`), and I just initialized them after calling `fix_collinearity` in the constructor. The problem here is that `fix_collinearity` uses `requires_graph`, which makes them entangled. I contemplated whether to have something like `collection.remove_metric(idx)` as you said, and update each time also things that can be just initialized later, or update only `requires_graph` which is necessary here. Obviously if we have more attributes to update, it will be better to export that logic to a different method, but I wasn't sure what is better for now where there is only a single such attribute.